### DISCUSSION TITLE
Using connection specific options in dialect methods of subclass from DDLCompiler

### DISCUSSION THREAD

**Discussion by hab6** [sqlalchemy/sqlalchemy#11574] [Usage Questions]
I am trying to update the Ingres [dialect](https://github.com/ActianCorp/sqlalchemy-ingres) code so that upon connecting to a target database, the DBMS catalog table [iidbcapabilities](https://docs.actian.com/actianx/12.0/index.html#page/DatabaseAdmin/Standard_Catalogs_for_All_Databases.htm#ww1029012) is queried to retrieve settings that can be associated with the connection and later used in method [IngresDDLCompiler::get_column_specification](https://github.com/ActianCorp/sqlalchemy-ingres/blob/262e6e817125d348d8d79c86e6bfc1339633113b/lib/sqlalchemy_ingres/base.py#L219).

The goal is for each connection object to have its own dictionary of catalog settings (retrieved once) that can be used for the life of the connection. This would allow for example the ability to adjust column clauses for the pending SQL `CREATE` statement appropriately for the backend (e.g. row-wise vs column-wise data store) and for it to work correctly when the front-end Python/SQLAlchemy application has multiple connections to different backends.

I read through the SQLAlchemy documentation on [sqlalchemy.engine.Connection.execution_options](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.execution_options) and tried various solutions, but can find no way in the `IngresDDLCompiler` class to get at the current connection object.

I must be missing something simple, but have yet to figure it out and appreciate any help.

**Comment by CaselIT** [comment]
Hi,

The connection is not passed to the compiler, you get only the dialect.
These kind of options are usually set on the dialect on first connect, in this method https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.Dialect.initialize

see for example the mysql one that does several queries to setup various configurations https://github.com/sqlalchemy/sqlalchemy/blob/c8d73b74733fa9f12b7b95a5885e808e6b31f1c6/lib/sqlalchemy/dialects/mysql/base.py#L2824-L2869

**Reply by hab6** [reply]
Yes, my prototype does just that, retrieving the catalog settings at dialect initialization. I tried storing a dictionary of these settings in `connection.execution_options`, but as you confirm, have no way to access the dictionary from the compiler class methods.

    class IngresDialect(default.DefaultDialect):
    ...
        def initialize(self, connection):
            super().initialize(connection)
            connection.execution_options(iidbcapabilities=self._load_iidbcapabilities(connection))

**Reply by CaselIT** [reply]
that would not be correct even if the compiler had a reference to the connection, since initialize is called just once per dialect instance, not once per connection, so you would have only one connection updated in any case.

**Reply by hab6** [reply]
Ah, that is helpful. Thanks for the clarification. I will need to consider another approach.

**Reply by CaselIT** [reply]
any reason for not setting these options on the dialect?

**Reply by hab6** [reply]
I have not really spent time understanding how SQLAlchemy handles multiple connections and have thought that a single instance of the dialect could have multiple connections to different data sources with each data source possibly have different catalog settings. If that is so, storing the catalog options in the dialect could be correct for one data source but not another. Maybe this thinking is incorrect?

**Reply by CaselIT** [reply]
are data sources similar to databases in a classic db? if that's the case in general you would have multiple instances of the dialect for each one.
I'm thinking along these lines:
```py
e1 = create_engine('ingress://.../ds1')
e2 = create_engine('ingress://.../ds2')
```

above e1 and e2 would have different instances of the dialect

**Reply by hab6** [reply]
I greatly appreciate the help and suggestions.

I have not yet gotten it working as desired by storing the catalog settings in the dialect, but I suspect there is a problem with my test case. Will continue debugging.

**Reply by hab6** [reply]
Thanks @CaselIT for helping me with this issue. Your comments allowed me to better understand how to approach a solution and also how to create a simple program with which to test it. I hope to soon merge changes to the Ingres dialect for loading database catalog settings that work in scenarios with multiple concurrent connections.
